Static Routes

Cydox opened the pull request:
This is the changes needed in common to facilitate adding a static route option to podman with the netavark backend.

containers/netavark#678
containers/netavark#680
containers/podman#18369

Cydox commented:
/hold

rhatdan commented:
@Luap99 PTAL

Luap99 commented on libnetwork/internal/util/parse.go:
```diff
@@ -34,3 +34,16 @@ func ParseVlan(vlan string) (int, error) {
+
+// ParseNoAutoGateway parses the no_auto_gateway option
+func ParseNoAutoGateway(opt string) (bool, error) {
+	switch opt {
```
you should use strconv.ParseBool() instead.

Cydox replied:
Done

Luap99 commented on libnetwork/netavark/config.go:
```diff
@@ -230,8 +235,14 @@ func (n *netavarkNetwork) networkCreate(newNetwork *types.Network, defaultNet bo
+	}
+
+	// add gateway when not internal or dns enabled and no_auto_gateway not set
+	addGateway := (!newNetwork.Internal || newNetwork.DNSEnabled) && (!noAutoGateway)
```
I just noticed but there is one problem with that, dns requires us to have a gateway ip so we can listen on it. Without it dns will not work. https://github.com/containers/common/commit/e030a76218e698f78bb9156b5c52f594f4c374d0

So we should either turn of DNS for this network if noAutoGw is set and throw a warning or just return an error and require users to create the bridge network with --disable-dns.

Cydox replied:
This is an issue indeed. Somebody could add a gateway manually if he uses no_auto_gateway though. So I think we should warn/error if the final config does not contain a gateway but the network doesn't have --disable-dns set.

I'm not certain what all the --disable-dns does, so I'm gonna do some digging in the code. Wanne make sure that if there is a route to a local dns server this can still be utilized if set in /etc/resolv.conf without a default gateway.

Cydox replied:
The solution now is to replace no_auto_gateway with the no_default_gateway option that prevents netavark from adding the default route but leaves the gateway in place. PR is containers/netavark#697

Cydox replied:
Pushed the changes here for no_default_gateway

Luap99 commented on libnetwork/types/network.go:
```diff
@@ -169,6 +171,17 @@ type Subnet struct {
+	// swagger:strfmt string
+	Gateway net.IP `json:"gateway"`
+	// Metric for this route. Optional.
+	Metric *int `json:"metric,omitempty"`
```
I think the metric type should be u32

Cydox replied:
Done

Luap99 reviewed (commented):
One last request, we need to validate the routes for invalid inputs, while we can agree that podman network create parses as a valid route we also accept a raw json payload via API so we need to ensure it is working. Right now I think there could be a bug where the gateway of the route is nil and netavark deserializer would not like that and error out. And the same for the Destination Ipnet, see the subnet validation code.

Also make sure to add tests in config_test.go to create networks with the new option and invalid values for the routes.

Cydox commented:
- Added code for validating that the routes gateway and destination are non-nil and that the destination is a network (instead of cidr address).
- Added tests for valid routes
- Added tests for invalid routes
- Added tests for no_default_route

Cydox commented:
Since containers/netavark#680 and containers/netavark#697 are merged I'll remove the hold on this.

/hold cancel

Cydox commented on libnetwork/netavark/config_test.go:
```diff
@@ -1356,6 +1356,325 @@ var _ = Describe("Config", func() {
+			},
+		}
+		_, err := libpodNet.NetworkCreate(network, nil)
+		Expect(err).To(HaveOccurred())
```
You're right! One minute...

Cydox replied:
Passed CI

Luap99 reviewed (approved):
LGTM
@mheon PTAL

mheon commented:
/lgtm
